### ISSUE TITLE
Make either representation polymorphic

### ISSUE THREAD

**Issue by treeowl** [haskell/core-libraries-committee#31]
Currently,

```haskell
either :: (a -> c) -> (b -> c) -> Either a b -> c
```

I propose we change this to the following:

```haskell
either :: forall {rep} a (c :: TYPE rep) b.
   (a -> c) -> (b -> c) -> Either a b -> c
```

Changing the (somewhat awkward) order of `a`, `b`, and `c` would break backwards compatibility; should we consider it anyway?

**Comment by Bodigrim** [comment]
Yeah, change of order is a breakage. I guess it depends whether you are up to prepare impact assessment or no.

**Comment by treeowl** [comment]
I am not up to preparing an impact assessment, though TBH I doubt there would be much breakage. I also don't have a real opinion about whether the order should be `a b c` or `c a b`. I think the type variable reordering question is orthogonal and can be discussed separately.

**Comment by andreasabel** [comment]
> I propose we change it to the following:
> 
> ```haskell
> either :: forall {rep} a (c :: TYPE rep) b.
>    (a -> c) -> (b -> c) -> Either a b -> c
> ```

(How) would that affect documentation?  I.e., how would that be rendered in the documentation of `either`?

Atm, I think that stuff like `forall {rep} (c :: TYPE rep)` isn't common-place Haskell...

**Comment by tomjaguarpaw** [comment]
See [`error`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:error) for comparison:

```haskell
error :: forall (r :: RuntimeRep). forall (a :: TYPE r). HasCallStack => [Char] -> a
```

Personally I find the leak of these details extremely unpleasant. Perhaps, in principle, the details could be hidden by Haddock though.

**Comment by andreasabel** [comment]
I also wonder whether this isn't a current short-coming of GHC?  Can anyone give some background on this new `RuntimeRep` thing?

It seems funny that GHC does not generalize type variable `a` in e.g.
```
error :: HasCallStack => String -> a`
```
to its _most general form_ (which seems now `forall {r} (a :: TYPE r)`.

(Compare this to Agda where you have to state how you want to generalize, so you would have to say either `variable a :: *` or `variable r :: RuntimeRep; a :: TYPE r`.)

**Comment by tbidne** [comment]
> I also wonder whether this isn't a current short-coming of GHC? Can anyone give some background on this new `RuntimeRep` thing?

@andreasabel  `RuntimeRep` is for [levity polymorphism](https://downloads.haskell.org/~ghc/9.0.2/docs/html/users_guide/exts/levity_polymorphism.html), i.e., polymorphism over boxed and unboxed types. The vast majority of haskell code is concerned with lifted types only, so there is no need for it. It is only useful for functions that should be defined for both boxed and unboxed types (e.g. [`($)`](https://hackage.haskell.org/package/base-4.16.0.0/docs/Prelude.html#v:-36-)).

**Comment by Bodigrim** [comment]
@treeowl would you like to proceed with this proposal? It awaits an impact assessment.

**Comment by Bodigrim** [comment]
@treeowl I assume no one is up to impact assessment of `either :: forall {rep} a b (c :: TYPE rep). ...` Would you like to put your original proposal with `either :: forall {rep} a (c :: TYPE rep) b. ...` for a vote?

**Comment by Bodigrim** [comment]
@treeowl if there is no further progress within two weeks, I'll close as abandoned.

**Comment by Bodigrim** [comment]
Closing as abandoned.
